### PR TITLE
[react-alert] Update .d.ts to v7 and add v4

### PR DESCRIPTION
**Pull request by g1eny0ung**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/schiehll/react-alert/releases I updated all types according to release notes and source code.
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/react-alert/index.d.ts [hunk 1]
```diff
@@ -1,22 +1,23 @@
-// Type definitions for react-alert 5.2
+// Type definitions for react-alert 7.0
 // Project: https://github.com/schiehll/react-alert
 // Definitions by: Yue Yang <https://github.com/g1eny0ung>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.8
```
Comment (peterblazejewicz) [inline types/react-alert/index.d.ts:5]:
```suggestion
```
this line is effectively dead code now, we could remove this IMO

File: types/react-alert/index.d.ts [hunk 2]
```diff
@@ -28,122 +29,127 @@ export interface Positions {
...
+export interface AlertOptions {
     /**
-     * The margin of each alert
+     * The margin of each alert.
      *
-     * Default value: '10px'
+     * Default: '10px'
      */
-    offset?: string | undefined;
+    offset?: string;
```
Comment (peterblazejewicz) [inline types/react-alert/index.d.ts:52]: here and in other places, those were added by DT maintainer to provide better interop with TS projects using exact optional property setting:
https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes
Those are optional in props:
https://github.com/schiehll/react-alert/blob/master/src/Provider.js#L176
this is optional

File: types/react-alert/index.d.ts [hunk 3]
```diff
@@ -1,22 +1,23 @@
...
-    | 'middle'
-    | 'middle right'
     | 'bottom left'
     | 'bottom center'
     | 'bottom right';
-
+export type AlertPosition = AlertPositionV4 | 'middle left' | 'middle' | 'middle right';
+export type AlertType = 'info' | 'success' | 'error';
+export type AlertTransition = 'fade' | 'scale';
 export interface Positions {
```
Comment (peterblazejewicz) [inline types/react-alert/index.d.ts:21]: just note, IMO those are actual symbols:
https://github.com/schiehll/react-alert/blob/master/src/options.js#L1
so in future we can rewrite this to `const`

**Comment by peterblazejewicz** [comment]
Are those old versions really required? Is there any use of them? (asking with concern of unneeded CI cycles to verify types on every `react` DT package update)
The v5 seems to be updated 3 years ago (as an update to NPM registry).
I'd say you could just stick to v7 and move it forward (previous versions of DT types remains published on NPM registry).

File: types/react-alert/index.d.ts (hunk 1 above)
Comment (g1eny0ung) [reply types/react-alert/index.d.ts:5]: Got it!

File: types/react-alert/index.d.ts (hunk 2 above)
Comment (g1eny0ung) [reply types/react-alert/index.d.ts:52]: Thanks for the reminder. I recognize the option now and will change these back later.

File: types/react-alert/index.d.ts (hunk 1 above)
Comment (g1eny0ung) [reply types/react-alert/index.d.ts]: Fixed in e18eb04.

File: types/react-alert/index.d.ts (hunk 2 above)
Comment (g1eny0ung) [reply types/react-alert/index.d.ts]: Fixed in d038e29.

**Comment by g1eny0ung** [comment]
> Are those old versions really required? Is there any use of them? (asking with concern of unneeded CI cycles to verify types on every `react` DT package update) The v5 seems to be updated 3 years ago (as an update to NPM registry). I'd say you could just stick to v7 and move it forward (previous versions of DT types remains published on NPM registry).

My initial intention was to have support for all versions, but I didn't realize the problem. 🤔 I agree with u. Perhaps we can remove `v3` and `v5` versions? Here are the reasons:

1. The `v2` implementation is different from the other versions, but now it still has 1500+ weekly downloads on npm, and this PR fixes some problems in the v2 version, so I hope it will be kept.
2. `v3` is more of a transitional version, the author rewrote the entire library in `v4` using the latest context API, so I think it could be removed.
3. `v5` supported hooks. But `v6` and `v7` just add changes to it, so it can be removed as well.

**Comment by g1eny0ung** [comment]
> > Are those old versions really required? Is there any use of them? (asking with concern of unneeded CI cycles to verify types on every `react` DT package update) The v5 seems to be updated 3 years ago (as an update to NPM registry). I'd say you could just stick to v7 and move it forward (previous versions of DT types remains published on NPM registry).
> 
> @peterblazejewicz My initial intention was to have support for all versions, but I didn't realize the problem. 🤔 I agree with u. Perhaps we can remove `v3` and `v5` versions? Here are the reasons:
> 
> 1. The `v2` implementation is different from the other versions, but now it still has 1500+ weekly downloads on npm, and this PR fixes some problems in the v2 version, so I hope it will be kept.
> 2. `v3` is more of a transitional version, the author rewrote the entire library in `v4` using the latest context API, so I think it could be removed.
> 3. `v5` supported hooks. But `v6` and `v7` just add changes to it, so it can be removed as well.

Finished in 00784df.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@g1eny0ung thx for accepting comments!

**Comment by g1eny0ung** [comment]
Ready to merge

**Pull request merged**